### PR TITLE
Vacation PR: multi-paged, styled, searchable docs with wrapper tooling and epubs for NixOS and Nixpkgs

### PR DESCRIPTION
**Pull request by grahamc**
TLDR:

 - New nixpkgs manual: http://gsc.io/70266391-48a6-49be-ab5d-acb5d7f17e76-nixpkgs/doc/nixpkgs-manual/html/index.html
 - Nixpkgs manual, epub: http://gsc.io/70266391-48a6-49be-ab5d-acb5d7f17e76-nixpkgs/doc/nixpkgs-manual/nixpkgs-manual.epub
 - New NixOS manual: http://gsc.io/70266391-48a6-49be-ab5d-acb5d7f17e76-nixos/doc/nixos-manual/html/index.html
 - NixOS manual, epub: http://gsc.io/70266391-48a6-49be-ab5d-acb5d7f17e76-nixos/doc/nixos-manual/nixos-manual.epub

###### Motivation for this change

 - The generated doc pages are huge and don't get indexed. 
 - Ctrl-F is a decent search tool, but a proper search engine with stemming is better.
 - Smaller pages are easier to navigate.
 - The output of `nix-build ` was boring.

###### Things done

1. Introduce a `docbook-index` tool for indexing DocBook against the generated HTML.
2. Split up the outputted documentation in to many smaller pages.
3. Added a whole new layout and theme to the docs which is consistent across NixOS and Nixpkgs.
4. Added a local, in-browser search engine (Elasticlunr, what Rust uses.)
5. Way reduced the complexity of generating the Nixpkgs and NixOS docs.
6. Made the Nixpkgs and NixOS docs build the same way.
7. Created `nix-doc-tools`, a metapackage used for building our documentation.
8. Updated the `how to write docs for nixos` docs to account for the `nix-doc-tools` tooling updates. 

BTW: nixpkgs and nixos docs are built the same way, so the docs apply to both.

Major thanks to @samueldr for the work on this.

- [x] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nix-review --run "nix-review wip"`
- [x] Tested execution of all binary files (usually in `./result/bin/`)
- [x] Ensured that relevant documentation is up to date
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

NOT DONE:
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
Need to determine the impact of building these docs on the minimal ISO.

I'm going on vacation, so I'll be back on the 16th or so to see what y'all think.

_btw, samueldr can make fixes to this branch for style/js things :)_

### CODE REVIEW

**Comment by jonringer** [comment]
manpages no longer exists on manual attr set:
https://github.com/NixOS/nixpkgs/blob/be8bd43041c834dcb47a11eb43d14cf225604ee2/nixos/modules/misc/documentation.nix#L194

I only see manual, manualHTML, manualHTMLShell
https://github.com/NixOS/nixpkgs/blob/be8bd43041c834dcb47a11eb43d14cf225604ee2/nixos/doc/manual/module.nix#L41

**Comment by veprbl** [comment]
Will this break all of the current URLs that refer to the anchors on the single page?

**Comment by samueldr** [comment]
> Will this break all of the current URLs that refer to the anchors on the single page?

Yes, as it is. I *think* we could work out a solution, literally thinking of it as I type.

 1. Prepare an index of anchors (they should still be named the same)
 1. On index.html, at load, cross-reference anchor with index, redirect.

The main issue there could be is history hijacking. It *may* be possible to do something tricky with html5 history, and window.reload, to overwrite the one entry.

###### Tasks

 * [ ] Implement JS for redirection on index
 * [ ] Implement anchors index

**Comment by davidak** [comment]
Great work!

Can we have the search box in the header like this?

Screenshot from 2019-09-04 18-38-28

**Comment by samueldr** [comment]
It might be possible, but it will cause issues with narrower layouts, especially with sections having longer titles.

There is another issue with the UX where the current implementation presents a clear separation of the search *mode* and the documentation consultation *mode*.

And additionally, while it looks like there is only a search box, there are elements in the search overlay that are currently unlikely to be seen. Among those is the loading state of the index. If we want to add more options to the search engine we would be furthermore constrained in space.

So, the short of it is that horizontal space is a premium, and it can't easily be spared for the search box. I did initially try it in the header, but it didn't feel right.

**Comment by grahamc** [comment]
_materializes from vacation for a few minutes_ hmm not a lot of feedback. I hope this can get in for 19.09! :) _disappears back to vacation_

**Comment by toonn** [comment]
Some feedback from someone who uses a narrow (for today's standards) browser. The index takes up a lot of horizontal space when you're not using it. [The python docs handle this really well](https://docs.python.org/3/library/re.html) by allowing you to fold the index.

Another thing that has always kinda annoyed me, the toc is super long (multiple accelerated scrolls long), and beginners have no need for it. The main page imo should either be the first section or a short explanation of the important bits of the manual, describing what's in the *nixpkgs* manual and that other manuals exist, e.g., nix and nixos manuals respectively and why they might want to read those instead (similar treatment for the other manuals).

**Comment by samueldr** [comment]
> Some feedback from someone who uses a narrow (for today's standards) browser. The index takes up a lot of horizontal space when you're not using it. The python docs handle this really well by allowing you to fold the index.

I guess you're not going narrow enough :)

image

I do use a narrow column, often, and have chosen what felt like an appropriate breakpoint to put the menu away. Though, I do agree that (maybe as a later step) a control could be added to toggle the menu at any levels.


###### Tasks

 * [ ] Index collapse/open control

**Review by jtojnar — COMMENTED** [review]
Big :+1: for fixing https://github.com/NixOS/nixpkgs/issues/68199 even before I opened it.

The tools look great, the only pain point is that `docs-build` is somewhat slower than `make out/html/index.html`. Maybe it would make sense to split the HTML and EPUB builds. I dream of fast rebuilds with live reloading the HTML page but I doubt there are free XSLT processors that support incremental builds.

**Comment by toonn** [comment]
> I guess you're not going narrow enough : )
@samueldr apparently it's about a cm's worth of difference. But when you go narrower you lose the index completely? The nice thing about the python docs is that you can *choose* whether you want the index visible or not and that it's never inaccessible.

File: pkgs/tools/misc/nix-doc-tools/default.nix [hunk 1]
```diff
@@ -27,15 +27,31 @@ let
...
+    set -eux
+
+    scratch=$(mktemp -d -t tmp.XXXXXXXXXX)
+    function finish {
+      rm -rf "$scratch"
+    }
+    trap finish EXIT
+
+    input="${root-file-name}"
+    output="$1"
```
[the remark this answers is not in the document]
Comment (samueldr) [reply pkgs/tools/misc/nix-doc-tools/default.nix]: As long as graham is fine with the change, a message has been added.

 * https://github.com/NixOS/nixpkgs/commit/d52120bc05ceee022a903432bd8a4e2964fb2a2e

**Comment by samueldr** [comment]
I have marked a bunch of @jtojnar's suggestions as resolved, they were trivial changes or changes I approve, in this WIP branch.

 * https://github.com/grahamc/nixpkgs/compare/samuel-docs/searchable-2...samueldr:feature/docs/searchable-3

There are additional changes in there that I would prefer graham would take a look first :)

**Comment by samueldr** [comment]
Answering @toonn's question:

> But when you go narrower you lose the index completely?

Not at all! There are three ways the narrow view allows the user to navigate:

  * The navigation icons at the top, to move to siblings or parent nodes.
  * The local TOC gets expanded at the top of the content, for easy visualization of the parent nodes, and the document's nodes.
  * The full TOC is present at the end of the document.

In fact, the narrow view shows more to the user, compared to the full view. Though, in reality the additional navigation at the end of the page is not as trivial to use.

> The nice thing about the python docs is that you can choose whether you want the index visible or not and that it's never inaccessible.

And you're right, it's nice to have the option to toggle it :).

**Comment by samueldr** [comment]
@toonn → https://stuff.samueldr.com/nixpkgs/manual/

 * :heavy_check_mark: That version has the togglable TOC.
 * :heavy_check_mark: That version has redirection.

To test the redirection:

 * https://stuff.samueldr.com/nixpkgs/manual/index.html#sec-emacs
 * https://stuff.samueldr.com/nixpkgs/manual/#sec-emacs

**Both** index.html and index.html-less  versions should redirect to the right location. That is because both index.html and index.html-less versions of the page are served by the nixos.org server.

I made sure to host the page at a location mimicking the URL structure from nixos.org.

**Comment by toonn** [comment]
@samueldr, very nice. All I'm missing is a friendlier "main" page than a detailed TOC like I mentioned earlier and possibly having toggles in the sidebar to open up items and dig down to sub sections. But this is already a major improvement imo. 👍

**Comment by samueldr** [comment]
> All I'm missing is a friendlier "main" page

This is out of scope of the PR; contents (except contribution to the documentation) is not part of this PR, considering this reviews the way docs function, both building and using documentation. Though I agree that this is a desirable improvement, to have something more on the main page.

> [...] having toggles in the sidebar to open up items and dig down to sub sections.

It was tried, but does not work well. Every level of the documentation has a page it links to. This meant that a control only used to toggle the open state of a section in the TOC would have to be implemented. Instead, I deemed it was more user-friendly to simply have one action for the options in the menu "follow and expand". With a local file, the navigation **is** instant, meaning there is no drawback. With hosted files (e.g. nixos.org) the navigation is near instant, due to the major decrease in page size. The cost in maintaining an additional non-native feature to the documentation is imho not worth the ability the collapse/expand TOC sections at will. That is especially true considering that following a link loses the expanded/collapsed state.

**Comment by toonn** [comment]
Then perhaps a static extra level to the TOC? The current sections are too broad imo, making the sidebar less useful.

File: doc/Makefile [hunk 2]
```diff
@@ -1,90 +1,7 @@
...
-
-functions/library/locations.xml: doc-support/result
-	ln -rfs ./doc-support/result/function-locations.xml functions/library/locations.xml
-
-functions/library/generated: doc-support/result
-	ln -rfs ./doc-support/result/function-docs functions/library/generated
+all: ${MD_TARGETS}
 
 %.section.xml: %.section.md
 	pandoc $^ -w docbook+smart \
```
Comment (jtojnar) [inline doc/Makefile]: According to the manual pandoc does not support smart extension for docbook.

https://pandoc.org/MANUAL#typography

**Comment by grahamc** [comment]
- [ ] `  <xsl:template name="make.toc">` is way too slow right now. Disabled for development, but needs to be re-enabled so the ToCs work properly.

**Comment by Mic92** [comment]
@grahamc is it known why xslt is too slow. Could be there a switch that only generates the expensive variant for nixos.org and not for local use?

**Comment by Mic92** [comment]
Should we close this PR?

**Comment by jonringer** [comment]
It does have a large number of conflicts. And not sure what direction the docs are going after the recent rfc and marketing team changes

cc @garbas

**Comment by jonringer** [comment]
@grahamc  your time and effort is still appreciated. Thank you for your hard work :)

**Comment by grahamc** [comment]
:) it was useful to me even if for no other reason than to convince me docbook isn’t worth the fight :).
